### PR TITLE
useNavigateRegions: Use closest to determine the next region to navigate to

### PR DESCRIPTION
**Pull request by alexstine**
## What?

This fixes a small bug in region navigation.

## Why?

Better accessibility.

## How?

Replaced:
```
		const selectedIndex = regions.indexOf(
			ref.current.ownerDocument.activeElement
		);
```
This only works if the current element is the region itself so if the user so much as interacts with the content within a region, the user will automatically start at region index 0 or the first region. This is super bad UX. My solution brings us closer to the way it should act with most screen readers but we're still off a bit.
With:
```
		const selectedIndex = regions.indexOf(
			ref.current.ownerDocument.activeElement.closest( '[role="region"]' )
		);
```
By using closest, we'll traverse upwards in the HTML until the wrapping region is found. This gives us the index in the regions array so we can best determine previous/next index.

## Testing Instructions

The following must be tested with a screen reader enabled such as NVDA. I did my testing in Firefox.
1. Open the editor.
2. Press tab to focus the View Posts link.
3. Press ctrl+` until you land on the Open document settings button.
4. Press tab a couple times.
5. Notice how the region navigation is based off where focus is currently placed. This is how screen readers work in-practice, now it is how Gutenberg works.

### CODE REVIEW

**Comment by ciampo** [comment]
Maybe @afercia could help with accessibility feedback here?

Or really anyone with a Windows + NVDA setup (I don't currently have one)

**Comment by afercia** [comment]
Thanks for the ping.
I'll try to test this on Windows, though I'll need to update my local environment which is a bit... old. I'm afraid that will take some time.

However, I'm not sure what kind of problem we're trying to solve.

useNavigateRegions was designed to emulate screen readers native landmarks navigation. It's not primarily designed for screen reader users. Instead, it aims to provide keyboard users who don't use a screen reader a navigation tool similar to the screen readers native one.

Worth reminding all screen readers provide landmarks navigation via dedicated shortcuts, for example:
- NVDA uses the D key.
- VoiceOver provides landmark navigation via the 'rotor'.
- JAWS uses the R key (if I remember correctly).

Improvements are always welcome but at first glance seems to me it's important to keep useNavigateRegions as close as possible to the native screen readers feature and not introduce changes. Will test to check if there's anything broken in the current implementation (the focus style is for sure) and then test the proposed changes.

**Comment by ciampo** [comment]
> Improvements are always welcome but at first glance seems to me it's important to keep useNavigateRegions as close as possible to the native screen readers feature and not introduce changes.

This was my gut feeling too — I usually prefer not to interfere with standard browser/screen reader behavior when possible!

I'd be curious to hear more in details why these changes are being proposed in the first place, and if anyone else could also chime in to help assess the potential problem and its proposed solution.

**Comment by alexstine** [comment]
These changes are being proposed because the current behavior is not accurate. I have recorded a screen share with audio to show how and why this should be improved.

@afercia The reason why we should not rely on landmark navigation with screen readers is also explained. Gutenberg requires stuff the accessibility spec is still lacking on so I have found a work-around just like other apps including Slack, Teams, and many others. Having a centralized way to move around that is not screen reader specific is a win every time.

https://drive.google.com/file/d/1DKeUpy26nr-fNiXkEmTl4X1zFbgRzLVh/view?usp=sharing

Hope this helps.

I am still planning on refining this a bit further. Just have not had the time.

Thanks.

**Comment by alexstine** [comment]
Okay, this is a complete solution now.
1. No matter where you are in the editor, navigating next will focus the next tabbable item in the next region.
2. No matter where you are in the editor, navigating previous will focus the previous tabbable item in the previous region.
3. This follows the exact same pattern NVDA navigation does but this comes with the advantage of taking control of the screen reader mode switching.
4. I modeled this after Teams and Slack with ctrl+f6 in Teams and f6 in Slack. When pressing these keys, it finds the next region and places focus on the content within it so the screen reader knows to switch modes. This saves a step with users constantly having to switch modes, we're handling it for them.

I honestly do not expect most people will understand this, it is fairly high-level workings of how screen readers work. However, I do hope this gets further review because I believe this could impact Gutenberg for the better.

Thanks.

**Comment by ciampo** [comment]
As you mentioned, reviewing this PR properly requires the reviewer(s) to have deep familiarity with how a screen reader works — since I don't, I will keep an eye on this PR and facilitate the review process if needed but I don't think I can be the main reviewer.

A couple of things that I wanted to point out:

- It sounds like we're implementing a behaviour that is specific to NVDA — would folks using a different screen readers feel disoriented ?
- It also sounds like the behavior is modelled after MS Teams and Slack — similarly to the previous point, is it a good idea to implement a custom behavior for navigating regions, instead of just following the browser/screen reader standards?
- There are some related failures in the e2e tests

**Comment by alexstine** [comment]
@ciampo Answers.

> It sounds like we're implementing a behaviour that is specific to NVDA — would folks using a different screen readers feel disoriented ?

No, this is common pattern across all screen readers. NVDA and Voiceover are free, that is why I suggest testing with them.

> It also sounds like the behavior is modelled after MS Teams and Slack — similarly to the previous point, is it a good idea to implement a custom behavior for navigating regions, instead of just following the browser/screen reader standards?

What I am trying to explain is the current version does not follow such standards. Placing focus on a negative tabindex of a region is not how any other screen reader handles it. Slack and Teams got this right.

The E2E tests are beyond me right now, I just want to get some feedback started.

I will also give this a test with Voiceover and JAWS over the weekend.

The important take-away here is this does not follow standards, my PR makes it follow more closely to what those standards are. Now it can be the same for everyone, screen reader or not.

Thanks.

**Comment by alexstine** [comment]
@ggordon-vispero If you have some spare time, could you please sanity check this one for me? I believe this brings our region navigation more in-line with how screen readers work. Feel free to disagree.

You can test at: http://gutenberg.run
Enter PR number: 44883
Navigate to the next part of the editor: Shift+Alt+N
Navigate to the previous part of the editor: Shift+Alt+P

As mentioned earlier in the PR, I tried to replicate what Slack has been doing with F6. I really enjoy how it places focus on the first tabbable since it can assist in switching screen reader mode automatically.

Please let me know if you need more context.

Thanks.

**Comment by afercia** [comment]
@alexstine thanks for the video and the clarifications. I know that requires patience and some effort.

To my understanding, we can maybe split this in 4 main points:
1. NVDA reads the entire content of the focused region.
2. Move focus to the first focusable instead of to the region.
3. Cycle through regions starting from the current region.
4. Empty tab stops.

On a general note, as mentioned earlier, this feature was originally designed for sighted keyboard users. I'm not sure optimizing this feature for screen readers is the way to go. We need to cover diverse needs of a divers range of users. I do believe that, in the long term, education is way more effective. I get one of your concerns is that many screen reader users don't know how to use native landmarks navigation. It would be great to educate them by the means of some good help guide instead of implementing hyper-technical solutions. Native features are always better.

That said, I'm all for improvements. There are a few problems though.

## NVDA reads the entire content of the focused region.
I'm afraid this is more a NVDA quirk rather than something we should fix. JAWS doesn't exhibit this behavior. VoiceOver neither. They just announce the region aria-label, as expected. I think we should report upstream this issue to NVDA. To me, when a landmark region receives focus and the region is labelled, NVDA should read only the label. I made a simplified codepen with 3 different tests, to make testing with various screen readers easier at https://codepen.io/afercia/full/PoegWdR

##  Move focus to the first focusable instead of the region.
It's important to note that the screen readers native landmarks navigation doesn't actually move focus. It moves the virtual cursor. This matters because the virtual cursor can move to any element, even the non-focusable ones. That is: if a landmark starts with a heading, or a paragraph, or a list, etc., landmarks navigation jumps to the landmark, moves the virtual cursor to the first element and announces it. This native behavior can be easily tested on the W3C landmark example pages, for instance at https://www.w3.org/WAI/ARIA/apg/example-index/landmarks/region.html
Instead, the proposed approach in this PR moves focus to the first focusable element. This would be risky because it would skip any non-focusable content placed at the start of a region. I guess we don't want users to miss content.

It's not the only problem though. Turns out that setting and removing the tabindex attribute as proposed in this PR doesn't work well in Chrome. See the third test on the [codepen](https://codepen.io/afercia/full/PoegWdR). With Chrome, removing the tabindex triggers a focus loss both with NVDA and JAWS. Nothing is announced.

Lastly, there's a visual thing to take into consideration. The current `useNavigateRegions` feature is supposed to draw a thick blue outline around the region that receives focus. Note that this is partially broken right now and works well only on the top bar and on the footer (reported on https://github.com/WordPress/gutenberg/issues/21717 more than 2 years ago). We should fix it but, regardless, the outline is designed to help sighted keyboard users in understanding what region is the currently focused one. This would be lost if we change to focus the first focusable element. As a low vision user, I really struggle to see where focus is when I use regions navigation with the changes from this PR.

Overall, I wouldn't recommend to move focus to the first element. We shouldn't optimize for screen readers only. We should always try to help the broadest range of users. Screen reader users can use the native landmarks navigation: it works better and we should educate them on how to use it. 

## Cycle through regions starting from the current region.
This would be a great improvement and I'd be all for it. RIght now, region navigation starts again from the first region and that's not ideal. It makes totally sense to start from the next region.

## Empty tab stops.
I'm not sure the empty tab stops you mention in your video are related to the negative `-1` tabindex set on the regions. A negative tabindex shouldn't have any effect on the tab sequence, as it makes an element programmatically focusable but the element isn't included in the tab sequence. Instead, I noticed some empty tab stops before and after the currently selected block. Those tab stops are pretty confusing. I think they come from the component that manages focus on the selected element. I just haven't had time to investigate further but seems to me this implementation is less than ideal. Not related to the regions though.

**Comment by MarcoZehe** [comment]
With regards to screen reader users not knowing native landmark navigation: That's an education problem that needs to be sorted out elsewhere. Gutenberg isn't the only complex site users will encounter on the web, so if they become involved in the web even a little more than just reading one or two news articles per day, they ought to learn their screen reader basics proper.

The more prevalent problem is that Gutenberg is so much of an app nowadays much more than a website, that native landmark navigation with the virtual cursor might not be available in a given context, because the screen reader is in application/focus/forms mode. The screen reader landmark navigation is only available once the virtual cursor is active. VoiceOver is the only exception, because they don't know that distinction of modes for historic reasons. So that's why Slack and others have implemented keyboard shortcuts like F6 to move to different sections of the application quickly, without forcing screen reader users into virtual mode even when the current context doesn't call for it.

**Review by MarcoZehe — APPROVED** [review]
I am giving this a positive review for the concept.

**Comment by alexstine** [comment]
@afercia Wanted to make this a bit more clear.

> Instead, the proposed approach in this PR moves focus to the first focusable element. This would be risky because it would skip any non-focusable content placed at the start of a region.

If you read the code diff, I added a comment explaining this. If I cannot find the first tabbable, I go ahead and place focus on the region just like it works before my changes. Can you think of a situation in a Gutenberg region where there is text before a tabbable element?

Aside from that, I think the big picture idea was caught the best by @MarcoZehe . I am trying to find a way to make mode switching a non-issue in Windows screen readers. I agree that education needs to be improved, but keep in mind that gets to be a political issue as well as a societal issue. I am careful how much I talk about this in public but the main difference between people who use screen readers and us devs is some screen reader users know enough to barely get by. We know screen readers at a higher level as we're trying to push the limits of existing technology to do more.

I also do not like the idea of adding a dedicated shortcut for this just for screen reader users. This functionality is well replicated in most apps. Pressing F6 will move focus around most Windows apps even browsers such as Google Chrome and Firefox. Elements such as the address bar, tabs, etc.

What I am saying is I do not want to avoid this due to this changing the experience for sighted users. We can fix the styling issues, we seem to really have no way forward for screen reader users. Gutenberg educating screen reader users is simply not a way forward for reasons mentioned above.

Thanks.

**Comment by afercia** [comment]
Note: while testing this, I noticed the macOS keyboard shortcut to navigate backwards `Shift + Control + backtick` fails on macOS Firefox and Safari. PR to fix it: https://github.com/WordPress/gutenberg/pull/45019

**Comment by afercia** [comment]
@MarcoZehe thanks for your feedback.

> With regards to screen reader users not knowing native landmark navigation: That's an education problem that needs to be sorted out elsewhere. 

I kindly disagree. One of the goals of a project like WordPress.org, apart from spreading out open source culture, is education. By that, I mean education of both users and developers.

That said, no one is proposing to remove this feature. It's going to stay. I disagree about changing it in a way that would make the experience actually worse for some users. For example, as a low vision user, focus jumping to an element that is visually very far from where focus was before, with a very light barely noticeable focus indication, is very bad. Instead, focusing the region itself, with a thick outline as focus indicator is far better. We need to take into considerations divers needs and optimizing for only one assistive technology isn't a good approach for universal access, in my humble opinion.

Anyways, for some data about landmarks use we can refer to the latest (and the previous ones) Webaim survey at https://webaim.org/projects/screenreadersurvey9/#landmarks. While it shows that frequent use of landmarks has decreased over time, it also shows that lot of users know what landmarks are. 

As you mentioned, the Gutenberg navigate regions feature has the advantage that it works also with screen readers in focus mode. That's great and one more reason to keep this feature. However, focusing the first element is a no go for me, as it worsen the experience for many non-screen reader users.

@alexstine 

> Can you think of a situation in a Gutenberg region where there is text before a tabbable element?

In the Site editor, some sidebars do have some short text before the first focusable. For example, the Styles and the Navigation sidebars. More importantly, in the future new regions may have text or other non-interactive content at the top. We can't assume the first content in a region is always a focusable element.

> I also do not like the idea of adding a dedicated shortcut for this just for screen reader users.

I'd agree.

> I also do not like the idea of adding a dedicated shortcut for this just for screen reader users. This functionality is well replicated in most apps. Pressing F6 will move focus around most Windows apps even browsers

When implementing this feature we did try to use F6. Unfortunately, it conflicts with operating system native shortcuts.

> We can fix the styling issues, we seem to really have no way forward for screen reader users. 

Not sure about keeping focus on the regions themselves would be a real barrier for screen reader users. Also, a couple points seem to me a bit in contradiction:
- One one hand you're saying most screen reader users don't know how landmarks work.
- On the other hand, you're saying: let's make this feature work as native landmarks do, so users know how it works.

To me, keeping focus on the regions seems the best compromise for all users and I'd recommend to not change the current behavior.

File: packages/components/src/higher-order/navigate-regions/index.js [hunk 1]
```diff
@@ -54,7 +56,17 @@ export function useNavigateRegions( shortcuts = defaultShortcuts ) {
 			nextRegion = regions[ nextIndex ];
 		}
 
-		nextRegion.focus();
+		// Add the tabindex of -1 so tabbable elements can be located.
+		nextRegion.setAttribute( 'tabindex', '-1' );
+		// Find the next tabbable within the region.
+		const nextTabbable = focus.tabbable.findNext( nextRegion );
+		// Is there a next tabbable available? If not, focus the region instead.
```
Comment (afercia) [inline packages/components/src/higher-order/navigate-regions/index.js]: I'm afraid there's no agreement on this point. Please change to not alter the current behavior.

File: packages/components/src/higher-order/navigate-regions/index.js [hunk 2]
```diff
@@ -54,7 +56,17 @@ export function useNavigateRegions( shortcuts = defaultShortcuts ) {
...
+		const nextFocus = nextTabbable ? nextTabbable : nextRegion;
+		// Focus our chosen area, either the first tabbable within the region or the region itself.
+		nextFocus.focus();
+		// Remove the tabindex, no longer needed after focus.
+		nextRegion.removeAttribute( 'tabindex' );
```
Comment (afercia) [inline packages/components/src/higher-order/navigate-regions/index.js]: As mentioned in the comments, removing tabindex dynamically  breaks the feature on Chrome and screen readers don't announce anything.

**Review by afercia — DISMISSED** [review]

**Comment by alexstine** [comment]
@afercia Okay, so some thoughts.

I for one have always questioned how correct a survey of users really is considering the ones that would not know what landmarks are probably would not be taking that survey in the first place. However, let's move on from that point.

> One one hand you're saying most screen reader users don't know how landmarks work.
On the other hand, you're saying: let's make this feature work as native landmarks do, so users know how it works.

This comes with the added benefit that users no longer need to worry about mode switching as much. I would say 98% of Gutenberg contributors have no idea what a screen reader mode is and this is why contributors get confused between the modes of Gutenberg and the modes of a screen reader. This further gets complicated because the Gutenberg mode toggle key (escape) is the same key that will exit edit/focus mode of a screen reader. The whole situation is convoluted for screen reader users and contributors alike. Gutenberg is a dynamic app with most of the functionality relying on keyboard input being sent through the application and screen readers not being in browse/virtual mode. Side question, have you ever tried to use Gutenberg in virtual/browse mode? It is not possible due to the ever changing DOM structure and DOM order is not followed.

The whole point here, this is a step in the right direction of forcing screen readers to remain in the proper mode as focus makes this happen depending on what type of ARIA role it comes in contact with.

I am not saying we should break anything visually, but myself not having vision, I can't fix that. I can only speak for what I know to work better.

I am reverting the focus and tabindex changes.

What you have suggested with browse/virtual mode will never work well for users in Gutenberg because screen reader modes manually have to be changed too often by the end-user. Not a problem for Voiceover users so Windows users get the bad end of this deal.

There have been some suggestions about wrapping Gutenberg in an application role but without the support of the community behind me, it is never going to be something I could pull off myself. All kinds of things will need to be considered first. My goal is to make the editor work with minimal effort because I understand that we're just not meeting the mark right now.

Not sure where else to take this PR, so I'll edit it down to size and that'll be it.

Thanks.

File: packages/components/CHANGELOG.md [hunk 3]
```diff
@@ -19,6 +20,7 @@
 -   `Navigator`: refactor unit tests to TypeScript and to `user-event` ([#44970](https://github.com/WordPress/gutenberg/pull/44970)).
 -   `Navigator`: Refactor Storybook code to TypeScript and controls ([#44979](https://github.com/WordPress/gutenberg/pull/44979)).
 -   `withFocusReturn`: Refactor tests to `@testing-library/react` ([#45012](https://github.com/WordPress/gutenberg/pull/45012)).
+>>>>>>> f2729bf25b342129078da946e6d05ef767cfa707
```
Comment (ciampo) [inline packages/components/CHANGELOG.md]: merge artifact

Comment (alexstine) [reply packages/components/CHANGELOG.md]: Dang, hate it when I miss those. Thanks for spotting.

**Review by ellatrix — APPROVED** [review]
This is good. When focus is within a region, it should indeed select the next region and not the first region in the document. Thank you!

**Comment by alexstine** [comment]
Oh, why are the E2E tests failing? These are not even related to my PR. Hopefully they'll pass eventually.

**Comment by alexstine** [comment]
@ciampo Do you know why the E2E tests keep failing on this? They also fail locally but I have not touched any area of the code for the avatar block.

**Comment by ellatrix** [comment]
Did you try a rebase? That test seems to be passing for trunk currently.

**Comment by alexstine** [comment]
@ellatrix I tried a `git rebase origin/trunk`. I have not really done branch rebase before, hopefully this was the way.

**Comment by alexstine** [comment]
@youknowriad @kevin940726 Do you have any ideas why these tests are still failing? This is far beyond my ability to debug at this point.

**Comment by kevin940726** [comment]
The playwright test is failing because we didn't update them in `a11y-region-navigation.spec.js`. Perhaps @afercia would know better how to correct them? It seems to me that we just need to add a couple more of `page.keyboard.press( 'Control+`' )` but I'm not entirely sure. I think it would be nice if we could add some comments to those `press()` functions to state where we're currently focusing on.

**Comment by alexstine** [comment]
Thanks @kevin940726 ! I think I managed to fix this up.

**Comment by alexstine** [comment]
@kevin940726 Do you know why the main E2E tests are still failing? Playwright is resolved, but I can't tell if my changes are causing test failures or if they are just flaky test related.

**Comment by kevin940726** [comment]
I think it's the same issue, we need to update the test to reflect the changes made here in: https://github.com/WordPress/gutenberg/blob/4f25464808098c9a4e4efd3de2e42a863f53c8a9/packages/e2e-tests/specs/editor/various/block-hierarchy-navigation.test.js#L122

**Comment by alexstine** [comment]
@kevin940726 Thanks! Those logs are always so confusing, I missed it.

@afercia Could you please check this and approve so this fix can be merged in?

Thanks.

**Comment by youknowriad** [comment]
Hi @afercia I see you still have a blocking review here. Did the recent changes and discussion address your points raised above? What's remaining here?

**Comment by afercia** [comment]
Sorry I missed the pings. Thanks everyone for this cool improvement.

**Pull request merged**